pkg/webhook/resourcedistribution/validating/resourcedistribution_create_update_handler.go

```diff
@@ -0,0 +1,203 @@
+		// validate namespace name
+		for _, msg := range coreval.ValidateNamespaceName(namespace.Name, false) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("excludedNamespaces"), targets.ExcludedNamespaces, msg))
+		}
+		// validate conflict between IncludedNamespaces and ExcludedNamespaces
+		if _, ok := included[namespace.Name]; ok {
+			conflicted = append(conflicted, namespace.Name)
+		}
+	}
+	if len(conflicted) != 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath, targets, fmt.Sprintf("ambiguous targets because namespace %v is in both IncludedNamespaces.List and ExcludedNamesapces.List", conflicted)))
+	}
+	if len(forbidden) != 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("includedNamespaces"), targets.IncludedNamespaces, fmt.Sprintf("cannot distribute rsource to forbidden namespaces %v", ForbiddenNamespaces)))
+	}
+
+	// 2. validate targets.NamespaceLabelSelector
+	if _, err := metav1.LabelSelectorAsSelector(&targets.NamespaceLabelSelector); err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("namespaceLabelSelector"), targets.IncludedNamespaces, fmt.Sprintf("labelSelectorAsSelector error: %v", err)))
+	}
+
+	return
+}
+
+// validateResourceDistribution is an entrance to validate ResourceDistribution when creating and updating
+// (1). validate ResourceDistribution ObjectMeta
+// (2). validate ResourceDistribution Spec
+func (h *ResourceDistributionCreateUpdateHandler) validateResourceDistribution(obj, oldObj *appsv1alpha1.ResourceDistribution, fldPath *field.Path) (allErrs field.ErrorList) {
+	// 1. validate metadata
+	allErrs = apimachineryvalidation.ValidateObjectMeta(&obj.ObjectMeta, false, apimachineryvalidation.NameIsDNSSubdomain, field.NewPath("metadata"))
+
+	// 2. validate spec
+	return append(allErrs, h.validateResourceDistributionSpec(obj, oldObj, fldPath.Child("spec"))...)
+}
+
+// Handle handles admission requests.
+func (h *ResourceDistributionCreateUpdateHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	obj := &appsv1alpha1.ResourceDistribution{}
+	oldObj := &appsv1alpha1.ResourceDistribution{}
+
+	switch req.AdmissionRequest.Operation {
+	case admissionv1beta1.Create:
+		if err := h.Decoder.Decode(req, obj); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+		if allErrs := h.validateResourceDistribution(obj, nil, nil); len(allErrs) > 0 {
+			klog.V(3).Infof("all errors when creating: %v", allErrs)
+			return admission.Errored(http.StatusUnprocessableEntity, allErrs.ToAggregate())
+		}
+	case admissionv1beta1.Update:
+		if err := h.Decoder.Decode(req, obj); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+		if err := h.Decoder.DecodeRaw(req.AdmissionRequest.OldObject, oldObj); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+
+		if allErrs := h.validateResourceDistribution(obj, oldObj, nil); len(allErrs) != 0 {
```

Comment:
Isn't the following code a bit more concise?

err := h.Decoder.Decode(req, obj);
if admissionv1beta1.Update:
   h.Decoder.DecodeRaw(req.AdmissionRequest.OldObject, oldObj);
h.validateResourceDistribution(obj, oldObj, nil)

Reply:
yes, you r right.